### PR TITLE
Speed up triu_tril_kernel

### PR DESCRIPTION
**Pull request by yuantailing**
1. Batch Processing: Enhance kernel efficiency by having each thread handle multiple elements, reducing the frequency of offset calculations.
2. Inplace Operation Optimization: For inplace functions, eliminate unnecessary copying to enhance performance.

Up to 5x speed up compared to torch 2.1.1

# Benchmark
Test on NVIDIA RTX 3080, WSL, CUDA 12.1. Peak performance is recorded.

  | function | dtype | shape | k | torch 2.1.1 | this PR | speed up
-- | -- | -- | -- | -- | -- | -- | --
various   dtype |   |   |   |   |   |  
  | triu_ | int8 | [1, 3072, 3072] | 0 | 0.107 | 0.028 | 3.76x
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.108 | 0.029 | 3.79x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.114 | 0.045 | 2.52x
  | triu_ | float64 | [1, 3072, 3072] | 0 | 0.172 | 0.082 | 2.11x
  | triu | int8 | [1, 3072, 3072] | 0 | 0.111 | 0.056 | 2.00x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.108 | 0.049 | 2.22x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.116 | 0.091 | 1.27x
  | triu | float64 | [1, 3072, 3072] | 0 | 0.175 | 0.176 | 1.00x
various   shape |   |   |   |   |   |  
  | triu_ | float32 | [1, 8192, 8192] | 0 | 0.798 | 0.311 | 2.56x
  | triu_ | float32 | [4, 1024, 1024] | 0 | 0.054 | 0.023 | 2.37x
  | triu_ | float32 | [4, 1021, 1021] | 0 | 0.054 | 0.023 | 2.33x
  | triu_ | float32 | [256, 128, 256] | 0 | 0.111 | 0.038 | 2.92x
  | triu_ | float32 | [128, 257, 125] | 0 | 0.051 | 0.029 | 1.77x
  | triu_ | float32 | [20480, 16, 16] | 0 | 0.072 | 0.036 | 1.97x
  | triu | float32 | [1, 8192, 8192] | 0 | 0.797 | 0.611 | 1.31x
  | triu | float32 | [4, 1024, 1024] | 0 | 0.056 | 0.042 | 1.32x
  | triu | float32 | [4, 1021, 1021] | 0 | 0.058 | 0.044 | 1.32x
  | triu | float32 | [256, 128, 256] | 0 | 0.114 | 0.093 | 1.22x
  | triu | float32 | [128, 257, 125] | 0 | 0.051 | 0.036 | 1.43x
  | triu | float32 | [20480, 16, 16] | 0 | 0.075 | 0.061 | 1.23x
various dim |   |   |   |   |   |  
  | triu_ | float32 | [3072, 3072] | 0 | 0.093 | 0.037 | 2.49x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.114 | 0.045 | 2.52x
  | triu_ | float32 | [1, 1, 3072, 3072] | 0 | 0.138 | 0.053 | 2.60x
  | triu | float32 | [3072, 3072] | 0 | 0.097 | 0.091 | 1.07x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.116 | 0.091 | 1.27x
  | triu | float32 | [1, 1, 3072, 3072] | 0 | 0.140 | 0.090 | 1.55x
various k |   |   |   |   |   |   |  
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.108 | 0.029 | 3.79x
  | triu_ | float16 | [1, 3072, 3072] | 1536 | 0.103 | 0.042 | 2.44x
  | triu_ | float16 | [1, 3072, 3072] | -1536 | 0.114 | 0.020 | 5.68x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.108 | 0.049 | 2.22x
  | triu | float16 | [1, 3072, 3072] | 1536 | 0.104 | 0.039 | 2.65x
  | triu | float16 | [1, 3072, 3072] | -1536 | 0.115 | 0.058 | 2.00x

# Benchmark Code

```python3
import time
import torch


torch.manual_seed(42)


def timeit(f, run_times=1000):
    torch.cuda.synchronize()
    t1 = time.time()
    for _ in range(run_times):
        f()
    torch.cuda.synchronize()
    t2 = time.time()
    return (t2 - t1) / run_times


for dtype in [torch.int8, torch.float16, torch.float32, torch.float64]:
    for shape in [
        [1, 8192, 8192],
        [3072, 3072],
        [1, 3072, 3072],
        [1, 1, 3072, 3072],
        [4, 1024, 1024],
        [4, 1021, 1021],
        [256, 128, 256],
        [128, 257, 125],
        [20480, 16, 16],
    ]:
        for k in [0, shape[-1] // 2, -shape[-1] // 2]:
            a = torch.empty(shape, dtype=dtype, device="cuda")
            for _ in range(4):
                t_triu = timeit(lambda: a.triu(k))
                t_triu_ = timeit(lambda: a.triu_(k))
                t_clone = timeit(lambda: a.clone())
                print(dtype, shape, f"{k=}", f"triu_ {t_triu_ * 1000:.6f} ({t_triu_ / t_clone:.2f}xMemcpy)", f"triu {t_triu * 1000:.6f} ({t_triu / t_clone:.2f}xMemcpy)")

            a = torch.rand(shape, device="cuda")
            a = (a * 10).to(dtype)
            assert (a.triu(k) == a.cpu().triu(k).cuda()).all()
            assert (a.tril(k) == a.cpu().tril(k).cuda()).all()
            assert (a.clone().triu_(k) == a.triu(k)).all()
            assert (a.clone().tril_(k) == a.tril(k)).all()
```

cc @ptrblck

### CODE REVIEW

**Comment by janeyx99** [comment]
@ptrblck would you be a good reviewer for this PR?

**Comment by yuantailing** [comment]
Hello @janeyx99, it seems that the pull request hasn't been reviewed yet. Would you kindly consider assigning a reviewer? Thank you!

**Comment by janeyx99** [comment]
@yuantailing Thanks for the reminder, I've ping'd @ptrblck who will be adding a reviewer from the CUDA side.

File: aten/src/ATen/native/cuda/TriangularOps.cu [hunk 1]
```diff
@@ -19,33 +19,53 @@
...
+
 namespace at::native {
 
 // ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ triu/tril ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-template <typename scalar_t, typename IndexType, bool upper>
-C10_LAUNCH_BOUNDS_1(cuda::getApplyBlockSize())
+constexpr static int block_size = 128;
+
+template <typename scalar_t, typename IndexType, bool upper, int elements_per_thread, bool inplace>
```
Comment (eqy) [inline aten/src/ATen/native/cuda/TriangularOps.cu:39]: It looks like there is a good amount of additional specialization that we're adding here. Did you find that specializing on `inplace` made a large performance difference over just making this a runtime parameter?

(just curious as we want to balance absolute runtime perf against binary size)

Comment (yuantailing) [reply aten/src/ATen/native/cuda/TriangularOps.cu:39]: If "inplace" is used as a runtime parameter, the performance of `triu_` becomes 87%~95%, and the performance of `triu` becomes 98%~102%, with the exception of a few outliers..

Due to slight environmental differences (e.g., variations in background processes), I retested the performance of "inplace as template," and there may be a slight difference compared to the measurements taken three weeks ago.

  | function | dtype | shape | k | as template | as param | rel. perf
-- | -- | -- | -- | -- | -- | -- | --
various   dtype |   |   |   |   |   |  
  | triu_ | int8 | [1, 3072, 3072] | 0 | 0.028 | 0.028 | 0.98x
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.030 | 0.032 | 0.94x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.046 | 0.051 | 0.89x
  | triu_ | float64 | [1, 3072, 3072] | 0 | 0.083 | 0.088 | 0.95x
  | triu | int8 | [1, 3072, 3072] | 0 | 0.054 | 0.054 | 1.01x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.047 | 0.048 | 0.99x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.089 | 0.090 | 0.98x
  | triu | float64 | [1, 3072, 3072] | 0 | 0.173 | 0.173 | 1.00x
various   shape |   |   |   |   |   |  
  | triu_ | float32 | [1, 8192, 8192] | 0 | 0.309 | 0.346 | 0.89x
  | triu_ | float32 | [4, 1024, 1024] | 0 | 0.022 | 0.025 | 0.90x
  | triu_ | float32 | [4, 1021, 1021] | 0 | 0.023 | 0.026 | 0.90x
  | triu_ | float32 | [256, 128, 256] | 0 | 0.037 | 0.042 | 0.88x
  | triu_ | float32 | [128, 257, 125] | 0 | 0.029 | 0.033 | 0.87x
  | triu_ | float32 | [20480, 16, 16] | 0 | 0.036 | 0.041 | 0.87x
  | triu | float32 | [1, 8192, 8192] | 0 | 0.600 | 0.603 | 0.99x
  | triu | float32 | [4, 1024, 1024] | 0 | 0.042 | 0.042 | 1.01x
  | triu | float32 | [4, 1021, 1021] | 0 | 0.042 | 0.041 | 1.01x
  | triu | float32 | [256, 128, 256] | 0 | 0.093 | 0.093 | 1.00x
  | triu | float32 | [128, 257, 125] | 0 | 0.035 | 0.036 | 0.98x
  | triu | float32 | [20480, 16, 16] | 0 | 0.059 | 0.061 | 0.98x
various dim |   |   |   |   |   |  
  | triu_ | float32 | [3072, 3072] | 0 | 0.037 | 0.042 | 0.89x
  | triu_ | float32 | [1, 3072, 3072] | 0 | 0.046 | 0.051 | 0.89x
  | triu_ | float32 | [1, 1, 3072, 3072] | 0 | 0.053 | 0.059 | 0.89x
  | triu | float32 | [3072, 3072] | 0 | 0.088 | 0.088 | 1.00x
  | triu | float32 | [1, 3072, 3072] | 0 | 0.089 | 0.090 | 0.98x
  | triu | float32 | [1, 1, 3072, 3072] | 0 | 0.089 | 0.094 | 0.95x
various k |   |   |   |   |   |   |  
  | triu_ | float16 | [1, 3072, 3072] | 0 | 0.030 | 0.032 | 0.94x
  | triu_ | float16 | [1, 3072, 3072] | 1536 | 0.042 | 0.041 | 1.02x
  | triu_ | float16 | [1, 3072, 3072] | -1536 | 0.020 | 0.022 | 0.94x
  | triu | float16 | [1, 3072, 3072] | 0 | 0.047 | 0.048 | 0.99x
  | triu | float16 | [1, 3072, 3072] | 1536 | 0.039 | 0.042 | 0.94x
  | triu | float16 | [1, 3072, 3072] | -1536 | 0.058 | 0.057 | 1.01x

Comment (yuantailing) [reply aten/src/ATen/native/cuda/TriangularOps.cu:39]: Although a template parameter, `elements_per_thread`, has been added, it corresponds one-to-one with `scalar_t`, so it does not increase the number of template specializations.

**Review by eqy — APPROVED** [review]

**Comment by yuantailing** [comment]
@janeyx99 @ptrblck Hello, does this PR need further review?

**Comment by janeyx99** [comment]
Nope! This is ready for merge. (I'm adding the -r since it's an old PR)

@pytorchbot merge -r

**Review by janeyx99 — APPROVED** [review]

File: aten/src/ATen/native/cuda/TriangularOps.cu [hunk 2]
```diff
@@ -19,33 +19,53 @@
 
 #include <ATen/cuda/CUDAApplyUtils.cuh>
 
+#define BOOL_SWITCH(COND, CONST_NAME, ...)      \
```
Comment (janeyx99) [inline aten/src/ATen/native/cuda/TriangularOps.cu:22]: I do notice we already have a definition of this in aten/native/transformers/cuda/flash_attn/static_switch.h, I wonder if this could be abstracted to a higher level, that is a common ancestor of the 2. This can come in a later PR. cc @drisspg

Comment (drisspg) [reply aten/src/ATen/native/cuda/TriangularOps.cu:22]: yeah, somewhere in c10 util would likely be a good location

Comment (yuantailing) [reply aten/src/ATen/native/cuda/TriangularOps.cu:22]: @janeyx99 There is also a similar macro called `DISPATCH_BOOL` in mem_eff_attention/gemm_kernel_utils.h:

https://github.com/pytorch/pytorch/blob/fee96adde766c401aaaeb3fda3e5240e084885d7/aten/src/ATen/native/transformers/cuda/mem_eff_attention/gemm_kernel_utils.h#L31-L40

cc @drisspg
